pandas/core/frame.py

```diff
@@ -4843,11 +4851,9 @@ def _sanitize_column(self, value) -> ArrayLike:
         """
         self._ensure_valid_index(value)
 
-        # We can get there through isetitem with a DataFrame
-        # or through loc single_block_path
-        if isinstance(value, DataFrame):
-            return _reindex_for_setitem(value, self.index)
-        elif is_dict_like(value):
+        # Using a DataFrame would mean coercing values to one dtype
+        assert not isinstance(value, DataFrame)
```

Comment:
no other cases that can be affected by disallowing this here?

Reply:
None that I can think of or find. Isetitem iterates over the columns now. Don’t think that we can get here with a DataFrame now

Reply:
Thanks for checking.

Reply:
Not for this PR, but some things I saw while double-checking:

in isetitem isinstance(value, DataFrame) case should we check that len(loc) == len(value.columns)?
_iset_item is only called from _replace_columnwise, where we know we have a Series with matching index, might be simpler to skip sanitize_column?
Might make sense to refactor _set_item back into `__setitem__`, as the latter isn't that complicated and the former has a name really similar to a bunch of other names

Reply:
1. Yes we should probably raise if they don't match
2. Should be faster at least
3. Makes __setitem__ pretty long, but shouldn't really hurt?